pensieve/statistics.py

```diff
@@ -0,0 +1,160 @@
+import attr
+import cattr
+import mozanalysis.bayesian_stats.bayesian_bootstrap as mabsbb
+from google.cloud import bigquery
+from google.api_core.exceptions import NotFound
+from typing import Callable, Any, Dict, List, Tuple, Optional
+from pandas import DataFrame
+import pandas
+import numpy as np
+
+from pensieve.pre_treatment import PreTreatment, RemoveNulls
+
+
+@attr.s(auto_attribs=True)
+class StatisticResult:
+    """
+    Represents the resulting data after applying a statistic transformation
+    to metric data.
+    """
+
+    metric: str
+    statistic: str
+    parameter: float
+    label: str
+    ci_width: Optional[float] = 0.0
+    point: Optional[float] = 0.0
+    lower: Optional[float] = 0.0
+    upper: Optional[float] = 0.0
+
+    def with_ci(self, data: DataFrame, t: float, confidence_level: float) -> "StatisticResult":
+        """Calculate the confidence interval and update result."""
+        confidence_margin = 0.5 * (1.0 - confidence_level)
+        confidence_high = (0.0 + confidence_margin) * 100
+        confidence_low = (1.0 - confidence_margin) * 100
+        self.lower = t - np.percentile(data - t, confidence_low)
+        self.upper = t - np.percentile(data - t, confidence_high)
+        self.ci_width = confidence_level
+        self.point = t
+        return self
+
+    def with_point(self, point_value: float) -> "StatisticResult":
+        """Set provided value as point value result for statistic."""
+        self.point = point_value
+        return self
+
+
+@attr.s(auto_attribs=True)
+class StatisticResultCollection:
+    """
+    Represents a set of statistics result data.
+    """
+
+    data: List[StatisticResult] = []
+
+    def append(self, result: StatisticResult):
+        self.data.append(result)
+
+    def merge(self, result_collection: "StatisticResultCollection"):
+        self.data = self.data + result_collection.data
```

Comment:
bikeshed: if `data` is public, why not just let people call `.append` and `.extend`?